### PR TITLE
[go/codegen]  Support for external resource references and imports/aliases

### PR DESCRIPTION
**Pull request by viveklak**
Fixes https://github.com/pulumi/pulumi/issues/5794

We are generating additional types for resources to cover container variants (pointers, lists, maps etc.) of inputs/outputs for resources since we can't predict which ones might be externally referenced.

Changes should only result in new types being created. Still expect some churn on the final types as we do more E2E testing on the new types.

Its a large diff so split up into multiple commits, in order of most important to less.

1. https://github.com/pulumi/pulumi/commit/e949ef319602852d4e4ceb02f84b9f5d1461d52d contains the core codegen changes
2. https://github.com/pulumi/pulumi/commit/9943354af075e3959723d280820383dd2374500f expands testing to cover all generated files for go codegen
3. https://github.com/pulumi/pulumi/commit/a018b4dc8e62e2520b1cb1edc48ab94d231f1746 adds test data to support the external reference resources etc. (1 and 2 above)
4. https://github.com/pulumi/pulumi/commit/af23ed614bf9bd5efcc33492d184a3dc7d195a0f contains test changes to support test schema changes for external-resources.

### CODE REVIEW

**Review by justinvp — APPROVED** [review]
LGTM!

Would love for @pgavlin and @lblackstone to review the generated types (as I'm less familiar with how these are supposed to look for Go).

**Review by lblackstone — APPROVED** [review]
Changes LGTM, but I'm a bit concerned by the amount of extra types this is going to generate. We're already running into problems on some of the larger providers, and this appears to approximately double the number of generated types.

It might be worth verifying that we can still compile e.g., kubernetes or azure-nextgen with these changes in place.

File: pkg/codegen/go/gen.go [hunk 1]
```diff
@@ -511,7 +591,7 @@ func (pkg *pkgContext) genEnumType(w io.Writer, name string, enumType *schema.En
 	elementType := pkg.enumElementType(enumType.ElementType, false)
 	fmt.Fprintf(w, "type %s %s\n\n", name, elementType)
 
-	fmt.Fprintln(w, "const (")
+	_, _ = fmt.Fprintln(w, "const (")
```
[the remark this answers is not in the document]
Comment (viveklak) [reply pkg/codegen/go/gen.go:594]: Removed.

File: pkg/codegen/go/gen.go [hunk 2]
```diff
@@ -319,6 +323,58 @@ func (pkg *pkgContext) inputType(t schema.Type, optional bool) string {
...
+	}
+	return pkg.tokenToResource(t.Token)
+}
+
+// resolveObjectType resolves resource references in properties while
+// taking into account potential external resources. Returned type is
+// always marked as required. Caller should check if the property is
+// optional and convert the type to a pointer if necessary.
+func (pkg *pkgContext) resolveObjectType(t *schema.ObjectType) string {
+	if t.Package != nil && pkg.pkg != nil && t.Package != pkg.pkg {
```
[the remark this answers is not in the document]
Comment (viveklak) [reply pkg/codegen/go/gen.go:359]: Fixed

File: pkg/codegen/go/gen.go [hunk 3]
```diff
@@ -319,6 +323,58 @@ func (pkg *pkgContext) inputType(t schema.Type, optional bool) string {
 	return typ + "Input"
 }
 
+// resolveResourceType resolves resource references in properties while
+// taking into account potential external resources. Returned type is
+// always marked as required. Caller should check if the property is
+// optional and convert the type to a pointer if necessary.
+func (pkg *pkgContext) resolveResourceType(t *schema.ResourceType) string {
+	if t.Resource != nil && pkg.pkg != nil && t.Resource.Package != pkg.pkg {
```
[the remark this answers is not in the document]
Comment (viveklak) [reply pkg/codegen/go/gen.go:331]: Fixed.

**Comment by viveklak** [comment]
> Changes LGTM, but I'm a bit concerned by the amount of extra types this is going to generate. We're already running into problems on some of the larger providers, and this appears to approximately double the number of generated types.
> 
> It might be worth verifying that we can still compile e.g., kubernetes or azure-nextgen with these changes in place.

I have been running locally with the kubernetes, aws and eks SDK with the additional types. The Go SDK has a bunch of compilation issues because we never compiled it as part of CI. Tracked [here](https://github.com/pulumi/pulumi-azure-nextgen-provider/issues/353) and am actively working on this.

**Review by lblackstone — APPROVED** [review]

**Comment by viveklak** [comment]
@lblackstone Further to your concerns about the additional types being added, I have pushed  https://github.com/pulumi/pulumi/pull/6124/commits/b98bcd26b944748fe8f875ad6967af0141ff698a to make generating the additional resource types opt-in. It does mean we need to make a couple of quick change to `pulumi-aws`, `pulumi-kubernetes` and `pulumi-eks`. I will cut a separate ticket to track eliminating its need. Since Go users are already inconvenienced by build times it might be best to be conservative here.

**Review by justinvp — APPROVED** [review]
I left a comment on https://github.com/pulumi/pulumi/commit/b98bcd26b944748fe8f875ad6967af0141ff698a

Just one question about how the option is specified. LGTM otherwise.

**Comment by viveklak** [comment]
> I left a comment on [b98bcd2](https://github.com/pulumi/pulumi/commit/b98bcd26b944748fe8f875ad6967af0141ff698a)
> 
> Just one question about how the option is specified. LGTM otherwise.

Updated now in https://github.com/pulumi/pulumi/pull/6124/commits/4da8a45eb2224ea09461388d79eb830286ea17fc

**Review by justinvp — APPROVED** [review]
LGTM

**Pull request merged**